decentraland/unity-explorer
# Fix: Make Archipelago flow truly recoverable

The previous flow:
- Didn't really use the auto-reconnection flow, that flow itself was a bit flawed
- `StartListeningForConnectionStringAsync` could be interrupted and never re-launched again, `catch` was outside of the loop
- There was a mix of `Result` and `Exceptions`: see [doc](https://github.com/decentraland/unity-explorer/wiki/Architecture-Overview#exceptions-free-async-flow)

Solution:
* Make the auto-reconnection flow be resistent to exceptions and interrupt only on cancellation. On Cancellation the flow will be re-launched from the connective room itself 
* Align and secure with `Result`/`LightResult` to exclude exceptions from the flow
* Enforce AutoReconnection
* Cover by unit tests

Technical thoughts, left for later
* `Archipelago` flow can be simplified: currently, it's broken down into 3 independent layers, each of them is represented by a series of decorators. It feels unjustifiedly complicated and quite hard to follow.
* There are multiple ways to achieve an exception-free flow: `Result`, `LightResult`, and `EnumResult<TErrorEnum>`. There could be a more generalized and clarified way of achieving essentially the same

Fixes #2686, #2447

## How to test the changes?

- Test the whole flow of Islands change: jumping between coordinates in Genesis City, and worlds
- Try to turn off your internet connection while `Island`:
   - Is being connected to (not connected yet)
   - Already connected
   - During teleportation (?)
- `Island` should recover when `Internet` is back in every case
